sklearn/tests/test_common.py

```diff
@@ -447,3 +447,51 @@ def test_estimators_do_not_raise_errors_in_init_or_set_params(Estimator):
 
         # Also do does not raise
         est.set_params(**new_params)
+
+
+ESTIMATOR_PARAMETER_TO_XFAIL = {
+    ("SpectralEmbedding", "affinity"),
+    ("FeatureUnion", "transformer_list"),  # Expected behavior
+    ("GaussianMixture", "weights_init"),
+    ("GaussianProcessRegressor", "alpha"),
+    ("LabelPropagation", "nn_fit"),
+    ("LabelSpreading", "nn_fit"),
+    ("NeighborhoodComponentsAnalysis", "n_components"),
+    ("PatchExtractor", "random_state"),
+    ("Pipeline", "steps"),  # Expected behavior
+    ("RANSACRegressor", "estimator"),  # Expected behavior
+    ("RidgeClassifier", "max_iter"),
+    ("RidgeClassifierCV", "alphas"),
+}
+
+
+@pytest.mark.parametrize("name, Estimator", all_estimators())
+def test_estimators_do_not_set_parameter_outside_of_init(name, Estimator):
+    """Check that parameter is not set outside of __init__."""
+    functions = chain.from_iterable(
+        (
+            f
+            for name, f in getmembers(kls, isfunction)
+            if f.__qualname__.startswith(kls.__name__) and name != "__init__"
+        )
+        for kls in Estimator.mro()
+    )
+
+    # Regex for "self.a_string = " where "a_string" can not begin or end with "_"
```

Comment:
naive question: is it not possible to find all "self.a_string" where "a_string" is in self.get_params() ?

Reply:
That would mostly work. The regex approach has the upside of catching when new parameter is set outside of `__init__`.